### PR TITLE
Change form validation on recurring contributions payment pages

### PR DESCRIPTION
**Pull request by jranks123**
## Why are you doing this?
We have recently been creating guest accounts without a first name and last name. This means that when a recurring contributor with a guest account signs in, they hit the payment page without a first name and last name. Previously, the would have hit this screen:

It is bad UX to hit a user with a red error message when they haven't done anything yet, so now we grey out the buttons instead, as well as adding the copy`All fields are required.` to the `Your Details` section. 

# Screenshots

## Control recurring page

| Before |
|--------|
| Valid |
| image |
| Invalid | 
| image |

| After |
|--------|
|Valid |
| image |
| Invalid | 
| image |

## Stripe inline payment recurring page

| Before |
|--------|
| Valid |
| image|
| Invalid | 
| image |

| After |
|--------|
|Valid |
| image |
| Invalid | 
| image |

## Things to check once merged

Check that the new validation works for both variants in the Stripe Inline test
Check that Stripe and Paypal payments can still be processed

### CODE REVIEW

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx [hunk 1]
```diff
@@ -59,12 +62,25 @@ function Button(props: PropTypes) {
 
   const PayPalButton = window.paypal.Button.driver('react', { React, ReactDOM });
 
+  if (props.disable) {
+    return (
+      <button
+        id="qa-pay-with-direct-debit"
```
Comment (svillafe) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx]: `qa-pay-with-direct-debit` I don't think this is the correct id

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx [hunk 2]
```diff
@@ -59,12 +62,25 @@ function Button(props: PropTypes) {
...
+        className={classNameWithModifiers('component-paypal-fake-pop-up-button', ['disable'])}
```
Comment (svillafe) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx]: Are we using the disable modifier here? I don't find in the css a place where is the definition of that class with that modifier.

File: assets/components/paymentButtons/directDebitPopUpButton/directDebitPopUpButton.jsx [hunk 3]
```diff
@@ -63,26 +65,33 @@ const DirectDebitPopUpButton = (props: PropTypes) => (
...
+  return <Button openPopUp={props.openDirectDebitPopUp} disable={props.disable} />;
 
 }
 
-function Button(props: { openPopUp: () => void }) {
+function Button(props: {
+  openPopUp: () => void,
+  disable: boolean
+  }) {
+
```
Comment (svillafe) [inline assets/components/paymentButtons/directDebitPopUpButton/directDebitPopUpButton.jsx]:
```
const className: string = classNameWithModifiers('component-direct-debit-pop-up-button', [disableClass]);

<button
       id="qa-pay-with-direct-debit"	       id="qa-pay-with-direct-debit"
       className="className"	      
className={disableClass}
 onClick={props.openPopUp}
     disabled={getDisabled}
     
```

Comment (jranks123) [reply assets/components/paymentButtons/directDebitPopUpButton/directDebitPopUpButton.jsx]: 👍 done

File: assets/components/yourDetails/yourDetails.jsx [hunk 4]
```diff
@@ -24,6 +24,7 @@ export default function YourDetails(props: PropTypes) {
     <div className="component-your-details">
       <PageSection heading="Your details" headingChildren={<Signout />}>
         <DisplayName />
+        <p className="component-your-details-text">All fields are required.</p>
```
Comment (Ap0c) [inline assets/components/yourDetails/yourDetails.jsx]: This is a [BEM element](http://getbem.com/naming/), so `component-your-details__text`.

This copy will also appear on the one-off checkout, just checking that's what you intended?

File: assets/components/contributionsCheckout/contributionsCheckout.scss [hunk 5]
```diff
@@ -9,6 +9,13 @@
     margin-top: $gu-v-spacing * 3;
   }
 
+  .component-your-details-text {
```
Comment (Ap0c) [inline assets/components/contributionsCheckout/contributionsCheckout.scss]: This is a [BEM element](http://getbem.com/naming/), so `component-your-details__text`.

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx [hunk 6]
```diff
@@ -59,6 +61,17 @@ function Button(props: PropTypes) {
 
   const PayPalButton = window.paypal.Button.driver('react', { React, ReactDOM });
 
+  if (props.disable) {
+    return (
+      <button
+        className="component-paypal-fake-pop-up-button"
+        disabled="true"
```
Comment (Ap0c) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx]: I think this should be `disabled`, boolean attributes in HTML take no values.

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx [hunk 7]
```diff
@@ -59,6 +61,17 @@ function Button(props: PropTypes) {
 
   const PayPalButton = window.paypal.Button.driver('react', { React, ReactDOM });
 
+  if (props.disable) {
+    return (
+      <button
```
Comment (Ap0c) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx]: Could this be inside the `component-paypal-button-checkout` div from below, with a className of `component-paypal-button-checkout__fake-pop-up-button`?

File: assets/components/paymentButtons/directDebitPopUpButton/directDebitPopUpButton.scss [hunk 8]
```diff
@@ -28,3 +28,18 @@
     fill: #fff;
   }
 }
+
+.component-direct-debit-pop-up-button--disable {
+  background-color: gu-colour(garnett-neutral-4);
+  color: gu-colour(garnett-neutral-2);
+  cursor: auto;
+
+  svg {
```
Comment (Ap0c) [inline assets/components/paymentButtons/directDebitPopUpButton/directDebitPopUpButton.scss]: `.svg-arrow-right-straight`?

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.scss [hunk 9]
```diff
@@ -5,3 +5,30 @@
...
+  position: relative;
+  text-align: left;
+  width: 100%;
+  margin-top: $gu-v-spacing;
+
+  @include mq($from: mobileMedium) {
+    width: 304px;
+  }
+
+  svg {
```
Comment (Ap0c) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.scss]: `.svg-arrow-right-straight`?

File: assets/components/paymentButtons/stripePopUpButton/stripePopUpButton.scss [hunk 10]
```diff
@@ -29,3 +29,17 @@
     background-color: darken(gu-colour(news-garnett-highlight), 5%);
   }
 }
+
+.component-stripe-pop-up-button--disable {
+  background-color: gu-colour(garnett-neutral-4);
+  color: gu-colour(garnett-neutral-2);
+  cursor: auto;
+  svg {
```
Comment (Ap0c) [inline assets/components/paymentButtons/stripePopUpButton/stripePopUpButton.scss]: `.svg-credit-card`?

File: assets/components/stripeInlineForm/stripeInlineForm.jsx [hunk 11]
```diff
@@ -136,14 +140,32 @@ function checkoutForm(props: {
...
       </label>
       <ErrorMessage message={props.errorMessage} />
-      <button className="component-stripe-inline-form__submit-payment">Confirm card payment <SvgArrowRightStraight /></button>
+      <button
+        className={className(submitPaymentBaseClass)}
+        disabled={props.disable}
+      >
+        Confirm card payment
+        <SvgArrowRightStraight />
+      </button>
```
Comment (Ap0c) [inline assets/components/stripeInlineForm/stripeInlineForm.jsx]: Could we avoid inline styles and the extra logic needed to apply individual modifier classes by applying a modifier on the enclosing component?

```jsx
<form
  className={classNameWithModifiers('component-stripe-inline-form', props.disable ? ['disabled'] : [])}
  onSubmit={handleSubmit}
>
  ...
</form>
```

```scss
.component-stripe-inline-form--disabled {
  pointer-events: none;

  .component-stripe-inline-form__submit-payment {
    // Disabled styles.
  }

  .component-stripe-inline-form__submit-payment {
    // Disabled styles.
  }

}
```

File: assets/components/stripeInlineForm/stripeInlineForm.scss [hunk 12]
```diff
@@ -52,3 +65,17 @@
     left: calc(100% - 34px);
   }
 }
+
+.component-stripe-inline-form__submit-payment--disable {
+  background-color: gu-colour(garnett-neutral-4);
+  color: gu-colour(garnett-neutral-2);
+
+  svg {
```
Comment (Ap0c) [inline assets/components/stripeInlineForm/stripeInlineForm.scss]: `.svg-arrow-right-straight`?

File: assets/pages/regular-contributions/components/regularContributionsPayment.jsx [hunk 13]
```diff
@@ -119,9 +117,10 @@ function RegularContributionsPayment(props: PropTypes, context) {
     isPostDeploymentTestUser={props.isPostDeploymentTestUser}
     amount={props.amount}
     switchStatus={props.stripeSwitchStatus}
+    disable={props.disable}
   />);
 
-  let payPalButton = (<PayPalExpressButton
```
Comment (Ap0c) [inline assets/pages/regular-contributions/components/regularContributionsPayment.jsx]: 👍

File: assets/pages/regular-contributions/components/regularContributionsPayment.jsx [hunk 14]
```diff
@@ -98,10 +94,12 @@ function RegularContributionsPayment(props: PropTypes, context) {
           context.store.getState,
       )}
         switchStatus={props.directDebitSwitchStatus}
+        error={props.error}
```
Comment (Ap0c) [inline assets/pages/regular-contributions/components/regularContributionsPayment.jsx]: This doesn't exist in the `PropTypes` for this button does it?

File: assets/pages/regular-contributions/components/regularInlineContributionsPayment.jsx [hunk 15]
```diff
@@ -104,10 +100,12 @@ function RegularContributionsPayment(props: PropTypes, context) {
           context.store.getState,
         )}
         switchStatus={props.directDebitSwitchStatus}
+        error={props.error}
```
Comment (Ap0c) [inline assets/pages/regular-contributions/components/regularInlineContributionsPayment.jsx]: Same as above, this isn't in the `PropTypes`.

File: assets/components/yourDetails/yourDetails.jsx (hunk 4 above)
Comment (jranks123) [reply assets/components/yourDetails/yourDetails.jsx]: Ah good spot. It wasn't my intention, but all fields are required for one-off too, so it makes sense in the context of the page.

File: assets/components/stripeInlineForm/stripeInlineForm.jsx (hunk 11 above)
Comment (jranks123) [reply assets/components/stripeInlineForm/stripeInlineForm.jsx]: 👍

File: assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx [hunk 16]
```diff
@@ -57,14 +59,24 @@ function Button(props: PropTypes) {
     props.callback,
   );
 
-  const PayPalButton = window.paypal.Button.driver('react', { React, ReactDOM });
+  const disabledButton = (
+    <button
+      className="component-paypal-button-checkout__disabled-pop-up-button"
+      disabled
+    >
+      Pay with Paypal
```
Comment (Ap0c) [inline assets/components/paymentButtons/payPalExpressButton/payPalExpressButton.jsx]: "PayPal"

**Pull request merged**